ipa-core/src/protocol/context/mod.rs

```diff
@@ -333,38 +362,170 @@ impl<'a> Inner<'a> {
     }
 }
 
+/// Reshards the given set of shares across all shards available on the current helper. It uses the
+/// provided function to determine destination shards for each input row.
+///
+/// The resulting vector contains all shares sent to this shard by other shards and shares selected
+/// locally by `shard_picker` fn to stay on this shard.
+///
+/// Resharding opens one communication channel per peer (N^2 channels will be open in total,
+/// N per shard). Each channel stays open until the very last row is processed, then they are explicitly
+/// closed, even if nothing has been communicated between that pair.
+///
+/// ## Panics
+/// It does not panic
+///
+/// ## Errors
+/// If cross-shard communication fails
+pub async fn reshard<L, K, C, S>(
```

Comment:
I am curious why this ended up in the context module?

I worry a bit that shard location could leak information, which will depend on how this function is used. Gluing it together with the local shuffle (which is the only way it is used currently), might mitigate that risk.

Reply:
I foresee this function being used in OPRF and other parts where we need shards to redistribute the shares. I don't see anything that cannot be otherwise done manually by opening a shard channel and sending data there. 

I also worry about timing attacks as intra-helper communication is now exposed and I was thinking that we need to add some protection on network layer - not sure if anything can be done in protocol code to prevent that